docx_editor/track_changes.py

```
@@ -43,6 +64,113 @@ def __init__(self, editor: DocxXMLEditor):
+        Raises:
+            HashMismatchError: If any paragraph hash is stale (no edits applied)
+            ValueError: If any operation is missing required fields
+        """
+        if not operations:
+            return []
+
+        # Parse and validate all refs upfront
+        parsed: list[tuple[int, ParagraphRef, EditOperation]] = []
+        for i, op in enumerate(operations):
+            if not op.paragraph:
+                raise ValueError(f"Operation {i}: paragraph reference is required for batch mode")
```

Ruff (0.15.7)

[warning] Avoid specifying long messages outside the exception class

(TRY003)